openedx/openedx-platform
# fix: allow non-Elasticsearch search engines when reindexing courses [FC-0062]

## Description

https://github.com/openedx/edx-search/pull/164 introduced the MeilisearchEngine to replace Elasticsearch for Sumac and later.

However, the management command used to [set up the course indexes](https://github.com/overhangio/tutor/blob/dbb91cb38f4c9d46dac28092745cc8afdb2c1b0c/tutor/commands/jobs.py#L177) has Elasticsearch-engine specific code in it, which fails when run with the Meilisearch engine.

This PR only fixes the error so that this management command can be used when indexing courses on Meilisearch, but it _does not_ set up the Meilisearch indexes -- this will be handled by Tutor.

## Supporting information

Relates to: https://github.com/overhangio/tutor/pull/1141
Part of: https://github.com/openedx/modular-learning/issues/236
Private-ref: [FAL-3903](https://tasks.opencraft.com/browse/FAL-3903)

## Testing instructions

1. Set up your tutor dev env with either the branch from https://github.com/overhangio/tutor/pull/1141 or by installing and enabling the [Meilisearch plugin](https://github.com/open-craft/tutor-contrib-meilisearch)
2. Enable course search indexing, e.g. with this plugin:
   ```yaml
   name: courseware_search
   version: 0.1.0
   patches:
     openedx-common-settings: |
       SEARCH_ENGINE = "search.meilisearch.MeilisearchEngine"
     lms-env-features: |
       ENABLE_COURSEWARE_INDEX: true
       ENABLE_COURSEWARE_SEARCH: true
     cms-env-features: |
       ENABLE_COURSEWARE_INDEX: true
       ENABLE_COURSEWARE_SEARCH: true
   ```
3. Create the Meilisearch indexes:
   ```
   tutor dev run cms ./manage.py cms shell -c "from search.meilisearch import create_indexes; create_indexes()"
   ```
4. Run the index setup command, and confirm that all courses are indexed:
   ```
   tutor dev run cms ./manage.py cms reindex_course --all --setup
   tutor dev run cms ./manage.py cms reindex_course --all
   ```
5. Confirm that you can reindex a course's content in Studio.

## Deadline

ASAP -- also needs to be backported to Sumac.

## Other information

1. Originally proposed to run `create_indexes()` when using `--setup`, but decided not to pollute the script with more engine-specific logic, and instead let Tutor handle this step.
